File: sdks/python/example_jwt.py
#!/usr/bin/env python3
"""
C2M API Python SDK Example with JWT Authentication
"""

import c2m_api
from c2m_api.rest import ApiException
import requests
import time
from datetime import datetime, timedelta

class C2MClient:
    def __init__(self, client_id, client_secret, base_url="https://api.click2mail.com/v2"):
        self.client_id = client_id
        self.client_secret = client_secret
        self.base_url = base_url
        self.auth_url = base_url
        self.long_token = None
        self.short_token = None
        self.token_expiry = None
        
    def get_long_token(self):
        """Get long-term token from auth service"""
        response = requests.post(
            f"{self.auth_url}/auth/tokens/long",
            json={
                "grant_type": "client_credentials",
                "client_id": self.client_id,
                "client_secret": self.client_secret,
                "scopes": ["jobs:submit", "jobs:read", "templates:read"],
                "ttl_seconds": 2592000  # 30 days
            }
        )
        response.raise_for_status()
        data = response.json()
        self.long_token = data["access_token"]
        print(f"Long-term token obtained: {data['token_id']}")
        return self.long_token
        
    def get_short_token(self):
        """Exchange long token for short-term token"""
        if not self.long_token:
            self.get_long_token()
            
        response = requests.post(
            f"{self.auth_url}/auth/tokens/short",
            headers={"Authorization": f"Bearer {self.long_token}"},
            json={"scopes": ["jobs:submit"]}
        )
        response.raise_for_status()
        data = response.json()
        self.short_token = data["access_token"]
        self.token_expiry = datetime.fromisoformat(data["expires_at"].replace('Z', '+00:00'))
        print(f"Short-term token obtained: {data['token_id']}")
        return self.short_token
        
    def ensure_authenticated(self):
        """Ensure we have a valid token"""
        if not self.short_token or datetime.now() >= self.token_expiry - timedelta(minutes=1):
            self.get_short_token()
        return self.short_token
        
    def create_api_client(self):
        """Create configured API client"""
        token = self.ensure_authenticated()
        
        configuration = c2m_api.Configuration()
        configuration.host = self.base_url
        configuration.api_key['bearerAuth'] = token
        configuration.api_key_prefix['bearerAuth'] = 'Bearer'
        
        return c2m_api.ApiClient(configuration)
    
    def submit_job(self, job_params):
        """Submit a job using the SDK"""
        with self.create_api_client() as api_client:
            jobs_api = c2m_api.DefaultApi(api_client)
            
            try:
                response = jobs_api.submit_single_doc_with_template_params(job_params)
                return response
            except ApiException as e:
                if e.status in [401, 403]:
                    # Token expired, refresh and retry
                    self.short_token = None
                    with self.create_api_client() as api_client:
                        jobs_api = c2m_api.DefaultApi(api_client)
                        response = jobs_api.submit_single_doc_with_template_params(job_params)
                        return response
                raise

# Usage example
if __name__ == "__main__":
    # Initialize client
    client = C2MClient(
        client_id="your-client-id",
        client_secret="your-client-secret"
    )
    
    # Submit a job
    job_params = {
        "templateId": "standard-letter",
        "documentUrl": "https://example.com/document.pdf",
        "recipients": [{
            "name": "John Doe",
            "address": {
                "line1": "123 Main St",
                "city": "New York",
                "state": "NY",
                "zip": "10001"
            }
        }]
    }
    
    try:
        result = client.submit_job(job_params)
        print(f"Job submitted: {result}")
    except Exception as e:
        print(f"Error: {e}")

File: sdks/javascript/example_jwt.js
/**
 * C2M API JavaScript SDK Example with JWT Authentication
 */

const C2mApi = require('c2m_api');
const axios = require('axios');

class C2MClient {
    constructor(clientId, clientSecret, baseUrl = 'https://api.click2mail.com/v2') {
        this.clientId = clientId;
        this.clientSecret = clientSecret;
        this.baseUrl = baseUrl;
        this.authUrl = baseUrl;
        this.longToken = null;
        this.shortToken = null;
        this.tokenExpiry = null;
    }
    
    async getLongToken() {
        const response = await axios.post(`${this.authUrl}/auth/tokens/long`, {
            grant_type: 'client_credentials',
            client_id: this.clientId,
            client_secret: this.clientSecret,
            scopes: ['jobs:submit', 'jobs:read', 'templates:read'],
            ttl_seconds: 2592000 // 30 days
        });
        
        this.longToken = response.data.access_token;
        console.log(`Long-term token obtained: ${response.data.token_id}`);
        return this.longToken;
    }
    
    async getShortToken() {
        if (!this.longToken) {
            await this.getLongToken();
        }
        
        const response = await axios.post(
            `${this.authUrl}/auth/tokens/short`,
            { scopes: ['jobs:submit'] },
            { headers: { Authorization: `Bearer ${this.longToken}` } }
        );
        
        this.shortToken = response.data.access_token;
        this.tokenExpiry = new Date(response.data.expires_at);
        console.log(`Short-term token obtained: ${response.data.token_id}`);
        return this.shortToken;
    }
    
    async ensureAuthenticated() {
        const now = new Date();
        const buffer = new Date(now.getTime() + 60000); // 1 minute buffer
        
        if (!this.shortToken || buffer >= this.tokenExpiry) {
            await this.getShortToken();
        }
        
        return this.shortToken;
    }
    
    async configureClient() {
        const token = await this.ensureAuthenticated();
        
        const defaultClient = C2mApi.ApiClient.instance;
        const bearerAuth = defaultClient.authentications['bearerAuth'];
        bearerAuth.accessToken = token;
        defaultClient.basePath = this.baseUrl;
        
        return defaultClient;
    }
    
    async submitJob(jobParams) {
        await this.configureClient();
        const jobsApi = new C2mApi.DefaultApi();
        
        try {
            return await jobsApi.submitSingleDocWithTemplateParams(jobParams);
        } catch (error) {
            if (error.status === 401 || error.status === 403) {
                // Token expired, refresh and retry
                this.shortToken = null;
                await this.configureClient();
                return await jobsApi.submitSingleDocWithTemplateParams(jobParams);
            }
            throw error;
        }
    }
}

// Usage example
async function main() {
    const client = new C2MClient(
        'your-client-id',
        'your-client-secret'
    );
    
    const jobParams = {
        templateId: 'standard-letter',
        documentUrl: 'https://example.com/document.pdf',
        recipients: [{
            name: 'John Doe',
            address: {
                line1: '123 Main St',
                city: 'New York',
                state: 'NY',
                zip: '10001'
            }
        }]
    };
    
    try {
        const result = await client.submitJob(jobParams);
        console.log('Job submitted:', result);
    } catch (error) {
        console.error('Error:', error);
    }
}

// Run if called directly
if (require.main === module) {
    main();
}

module.exports = C2MClient;
